src/server.c

```diff
@@ -4270,53 +4348,107 @@ void timeCommand(client *c) {
 }
 
 /* Helper function for addReplyCommand() to output flags. */
-int addReplyCommandFlag(client *c, struct redisCommand *cmd, int f, char *reply) {
-    if (cmd->flags & f) {
+int addReplyCommandFlag(client *c, int flags, int f, char *reply) {
+    if (flags & f) {
         addReplyStatus(c, reply);
         return 1;
     }
     return 0;
 }
 
+void addReplyFlagsForCommand(client *c, struct redisCommand *cmd) {
+    int flagcount = 0;
+    void *flaglen = addReplyDeferredLen(c);
+    flagcount += addReplyCommandFlag(c,cmd->flags,CMD_WRITE, "write");
+    flagcount += addReplyCommandFlag(c,cmd->flags,CMD_READONLY, "readonly");
+    flagcount += addReplyCommandFlag(c,cmd->flags,CMD_DENYOOM, "denyoom");
+    flagcount += addReplyCommandFlag(c,cmd->flags,CMD_ADMIN, "admin");
+    flagcount += addReplyCommandFlag(c,cmd->flags,CMD_PUBSUB, "pubsub");
+    flagcount += addReplyCommandFlag(c,cmd->flags,CMD_NOSCRIPT, "noscript");
+    flagcount += addReplyCommandFlag(c,cmd->flags,CMD_RANDOM, "random");
+    flagcount += addReplyCommandFlag(c,cmd->flags,CMD_SORT_FOR_SCRIPT,"sort_for_script");
+    flagcount += addReplyCommandFlag(c,cmd->flags,CMD_LOADING, "loading");
+    flagcount += addReplyCommandFlag(c,cmd->flags,CMD_STALE, "stale");
+    flagcount += addReplyCommandFlag(c,cmd->flags,CMD_SKIP_MONITOR, "skip_monitor");
+    flagcount += addReplyCommandFlag(c,cmd->flags,CMD_SKIP_SLOWLOG, "skip_slowlog");
+    flagcount += addReplyCommandFlag(c,cmd->flags,CMD_ASKING, "asking");
+    flagcount += addReplyCommandFlag(c,cmd->flags,CMD_FAST, "fast");
+    flagcount += addReplyCommandFlag(c,cmd->flags,CMD_NO_AUTH, "no_auth");
+    flagcount += addReplyCommandFlag(c,cmd->flags,CMD_MAY_REPLICATE, "may_replicate");
+    if (cmdHasMovableKeys(cmd)) {
+        addReplyStatus(c, "movablekeys");
+        flagcount += 1;
+    }
+    setDeferredSetLen(c, flaglen, flagcount);
+}
+
+void addReplyFlagsForKeyArgs(client *c, int flags) {
+    int flagcount = 0;
+    void *flaglen = addReplyDeferredLen(c);
+    flagcount += addReplyCommandFlag(c,flags,CMD_KEY_WRITE, "write");
+    flagcount += addReplyCommandFlag(c,flags,CMD_KEY_READ, "read");
+    setDeferredSetLen(c, flaglen, flagcount);
+}
+
+void addReplyCommandKeyArgs(client *c, struct redisCommand *cmd) {
+    int speccount = 0;
+    void *speclen = addReplyDeferredLen(c);
```

Comment:
can't we know the count in advance? I.e. there's no need for a deferred reply here.